core/src/main/java/com/graphhopper/storage/ShortcutUnpacker.java

```diff
@@ -31,60 +34,85 @@ public ShortcutUnpacker(Graph graph, Visitor visitor) {
         if (edge == null) {
             throw new IllegalArgumentException("Edge with id: " + edgeId + " does not exist or does not touch node " + adjNode);
         }
-        expandEdge(edge, false);
+        expandEdge(edge, reverse, prevOrNextEdgeId);
     }
 
-    private void expandEdge(CHEdgeIteratorState edge, boolean reverse) {
+    private void expandEdge(CHEdgeIteratorState edge, boolean reverse, int prevOrNextEdgeId) {
         if (!edge.isShortcut()) {
-            // todo: should properly pass previous edge here. for example this is important for turn cost time evaluation
-            // with edge-based CH, #1585
-            visitor.visit(edge, reverse, EdgeIterator.NO_EDGE);
+            visitor.visit(edge, reverse, prevOrNextEdgeId);
             return;
         }
-        expandSkippedEdges(edge.getSkippedEdge1(), edge.getSkippedEdge2(), edge.getBaseNode(), edge.getAdjNode(), reverse);
+        if (edgeBased) {
+            expandSkippedEdgesEdgeBased(edge.getSkippedEdge1(), edge.getSkippedEdge2(), edge.getBaseNode(), edge.getAdjNode(), reverse, prevOrNextEdgeId);
+        } else {
+            expandSkippedEdgesNodeBased(edge.getSkippedEdge1(), edge.getSkippedEdge2(), edge.getBaseNode(), edge.getAdjNode(), reverse);
+        }
     }
 
-    private void expandSkippedEdges(int skippedEdge1, int skippedEdge2, int from, int to, boolean reverse) {
-        // for edge-based CH we need to take special care for loop shortcuts
-        if (from == to) {
-            CHEdgeIteratorState sk1 = getEdge(skippedEdge1, from);
-            CHEdgeIteratorState sk2 = getEdge(skippedEdge2, from);
-            if (sk1.getAdjNode() == sk1.getBaseNode() || sk2.getAdjNode() == sk2.getBaseNode()) {
-                // this is a loop where both skipped edges are loops. but this should never happen.
-                throw new IllegalStateException(String.format(Locale.ROOT,
-                        "error: detected edge where both skipped edges are loops. from: %d, to: %d, " +
-                                "skip-edge1: %d, skip-edge2: %d, reverse: %b", from, to, skippedEdge1, skippedEdge2, reverse));
-            }
+    private void expandSkippedEdgesEdgeBased(int skippedEdge1, int skippedEdge2, int base, int adj, boolean reverse, int prevOrNextEdgeId) {
+        if (reverse) {
+            int tmp = skippedEdge1;
+            skippedEdge1 = skippedEdge2;
+            skippedEdge2 = tmp;
+        }
+        CHEdgeIteratorState sk2 = getEdge(skippedEdge2, adj);
+        assert sk2 != null : "skipped edge " + skippedEdge2 + " + is not attached to adjNode " + adj + ". this should " +
+                "never happen because edge-based CH does not use bidirectional shortcuts at the moment";
+        CHEdgeIteratorState sk1 = getEdge(skippedEdge1, sk2.getBaseNode());
+        if (base == adj && (sk1.getAdjNode() == sk1.getBaseNode() || sk2.getAdjNode() == sk2.getBaseNode())) {
+            throw new IllegalStateException(String.format(Locale.ROOT,
```

Comment:
Does this never happen because we remove loops in the import? Or this is something else as we still have loop support elsewhere like in getShortcutUnpacker?

Reply:
I think a skipped edge that is a loop is impossible unless there is a bug in CH preparation. I think this check has been there more or less since the very beginning and I think I never saw this error :)

Reply:
I did not see this error before and after we excluded loops during import.

Reply:
Regarding loop support I think our algorithms can handle them in most cases but what is not possible is to decide in which direction a loop should be read (problematic when there is geometry or different fwd/bwd speeds etc.)

Reply:
Ok, thanks for the explanation!

> (problematic when there is geometry or different fwd/bwd speeds etc.)

Should we create a separate issue for this? I think there is a solution (although not without problems) https://github.com/graphhopper/graphhopper/issues/1525#issuecomment-456138104

Reply:
Yes! And another one for some turn-cost cleanup stuff, like u-turn in calcMillis and calcTurnCost in Weighting. Will do.

Reply:
Thanks, so from my perspective you can merge (have not fully understood everything but is also not required yet ;) )

Reply:
Ok thanks for the review! The changes in shortcutunpacker and pathbidirref are really not easy to understand without thinking about it in detail, sketching some examples etc. Was not so easy to get it right to be honest. I might post a little drawing to make it clear.